src/workerd/api/node/crypto.c++

```diff
@@ -943,7 +944,262 @@ void CryptoImpl::CipherHandle::setAuthTag(jsg::Lock& js, jsg::BufferSource authT
 
 jsg::BufferSource CryptoImpl::CipherHandle::getAuthTag(jsg::Lock& js) {
   JSG_REQUIRE(!ctx, Error, "Auth tag is only available once cipher context has been finalized");
-  JSG_REQUIRE(mode == Mode::CIPHER, Error, "Getting the auth tag is only support for cipher");
+  JSG_REQUIRE(mode == CipherMode::CIPHER, Error, "Getting the auth tag is only support for cipher");
+
+  KJ_IF_SOME(tag, maybeAuthTag) {
+    KJ_DEFER(maybeAuthTag = kj::none);
+    return kj::mv(tag);
+  }
+
+  auto backing = jsg::BackingStore::alloc<v8::ArrayBuffer>(js, 0);
+  return jsg::BufferSource(js, kj::mv(backing));
+}
+
+namespace {
+CryptoImpl::AeadHandle::AuthenticatedInfo initAuthenticated(ncrypto::Aead& aead,
+    ncrypto::AeadCtxPointer& ctx,
+    bool encrypt,
+    kj::StringPtr cipher_type,
+    int iv_len,
+    unsigned int auth_tag_len) {
+  ncrypto::MarkPopErrorOnReturn mark_pop_error_on_return;
+
+  CryptoImpl::AeadHandle::AuthenticatedInfo info;
+  info.auth_tag_len = auth_tag_len;
+
+  const int mode = aead.getMode();
+  if (mode == EVP_CIPH_GCM_MODE) {
+    if (info.auth_tag_len != kNoAuthTagLength) {
+      JSG_REQUIRE(ncrypto::Cipher::IsValidGCMTagLength(auth_tag_len), Error,
+          "Invalid authentication tag length");
+    }
+  } else {
+    if (auth_tag_len == kNoAuthTagLength) {
+      // We treat ChaCha20-Poly1305 specially. Like GCM, the authentication tag
+      // length defaults to 16 bytes when encrypting. Unlike GCM, the
+      // authentication tag length also defaults to 16 bytes when decrypting,
+      // whereas GCM would accept any valid authentication tag length.
+      if (aead.getName() == "chacha20-poly1305"sv) {
+        info.auth_tag_len = 16;
+      } else {
+        JSG_FAIL_REQUIRE(
+            Error, kj::str("The auth tag length is required for cipher ", cipher_type));
+      }
+    }
+
+    if (mode == EVP_CIPH_CCM_MODE && !encrypt && FIPS_mode()) {
+      JSG_FAIL_REQUIRE(Error, "CCM encryption not supported in FIPS mode");
+    }
+
+    if (mode == EVP_CIPH_CCM_MODE) {
+      JSG_REQUIRE(iv_len >= 7 && iv_len <= 13, Error, "Invalid authentication tag length");
+      if (iv_len == 12) info.max_message_size = 16777215;
```

Comment:
can you add a comment for these magical numbers

Reply:
These are copy-pasted from CipherHandle, so I'm going to hereby bug @jasnell to explain this.

Reply:
These are specific standard requirements for CCM cipher IV lengths.